crypto/BRCryptoTransfer.c

```diff
@@ -89,9 +93,43 @@ cryptoTransferCreateAsBTC (BRCryptoCurrency currency,
                            BRWallet *wid,
                            BRTransaction *tid) {
     BRCryptoTransfer transfer = cryptoTransferCreateInternal (BLOCK_CHAIN_TYPE_BTC, currency);
-    transfer->u.btc.wid = wid;
     transfer->u.btc.tid = tid;
 
+    // cache the values that require the wallet
+    transfer->u.btc.fee  = BRWalletFeeForTx (wid, tid);
+    transfer->u.btc.recv = BRWalletAmountReceivedFromTx (wid, tid);
+    transfer->u.btc.send = BRWalletAmountSentByTx (wid, tid);
+
+    {
+        size_t     inputsCount = tid->inCount;
+        BRTxInput *inputs      = tid->inputs;
+
+        int inputsContain = (UINT64_MAX != transfer->u.btc.fee ? 1 : 0);
+
+        for (size_t index = 0; index < inputsCount; index++) {
+            if (inputsContain == BRWalletContainsAddress(wid, inputs[index].address)) {
+                BRAddress address;
+                memcpy (address.s, inputs[index].address, sizeof (address.s));
+                transfer->sourceAddress = cryptoAddressCreateAsBTC (address);
+            }
+        }
+    }
+
+    {
+        size_t      outputsCount = tid->outCount;
+        BRTxOutput *outputs      = tid->outputs;
+
+        int outputsContain = (UINT64_MAX == transfer->u.btc.fee ? 1 : 0);
+
+        for (size_t index = 0; index < outputsCount; index++) {
+            if (outputsContain == BRWalletContainsAddress(wid, outputs[index].address)) {
+                BRAddress address;
+                memcpy (address.s, outputs[index].address, sizeof (address.s));
+                transfer->targetAddress = cryptoAddressCreateAsBTC (address);
+            }
+        }
+    }
+
```

Comment:
There is no `break` in the above?  `transfer->targetAddress` is assigned possibly multiple times?  Ditto `transfer->sourceAddress` above?

Reply:
Good catch. Fixed.